### PR TITLE
ENH: Added stride/offset aliases in to_datetime

### PR DESCRIPTION
**Pull request by timcera**
Added support for stride and offset aliases to the 'unit' keyword in
'to_datetime'.  Documented support for additional 'unit' intervals that
were available.

- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by WillAyd — CHANGES_REQUESTED** [review]
Does this close an existing issue? Also tests are critical to make sure this actually works

**Review by mroeschke — CHANGES_REQUESTED** [review]
Yearly and monthly offsets were deprecated recently. https://github.com/pandas-dev/pandas/issues/16344. 

Week offsets are on the chopping block too so it's best not to advertise them as well. https://github.com/pandas-dev/pandas/issues/14024#issuecomment-240680919

**Comment by timcera** [comment]
> Does this close an existing issue? Also tests are critical to make sure this actually works

I couldn't find an existing issue.  I had an issue that I didn't report, however.  I was trying to use `to_datetime` to make an hourly series and blindly used `to_datetime([1, 2], unit='H')` didn't work, which was puzzling.  So looked at the code finding that `to_datetime([1, 2], unit='h')` did work, but 'h' is not an offset alias.  Other strange things where, 'm' meant minutely, and 'M' meant monthly.

After working with the code found it easy to add stride.

Passed current tests, so I didn't break old behaviors.  

Where should I put new tests?

**Comment by timcera** [comment]
> Yearly and monthly offsets were deprecated recently. #16344.
> 
> Week offsets are on the chopping block too so it's best not to advertise them as well. [#14024 (comment)](https://github.com/pandas-dev/pandas/issues/14024#issuecomment-240680919)

There isn't a deprecation warning in `to_datetime()`.  Will add deprecation warning for 'Y', 'A', 'M', and 'W'.

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 1]
```diff
@@ -252,35 +254,54 @@ cpdef inline object precision_from_unit(object unit):
...
+    # Normalize old or lowercase codes to standard offset aliases.
+    if unit in ['min', 'm']:
+        unit = 'T'
+    elif unit == 'ns':
+        unit = 'N'
+    elif unit == 'ms':
+        unit = 'L'
+    elif unit == 'us':
+        unit = 'U'
+
```
Comment (jreback) [inline pandas/_libs/tslibs/timedeltas.pyx]: make this an else

File: pandas/core/tools/datetimes.py [hunk 2]
```diff
@@ -470,11 +470,26 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
...
 
-    unit : string, default 'ns'
-        unit of the arg (D,s,ms,us,ns) denote the unit, which is an
-        integer or float number. This will be based off the origin.
-        Example, with unit='ms' and origin='unix' (the default), this
-        would calculate the number of milliseconds to the unix epoch start.
+    unit : string, default is 'N'
+        The unit of the arg. Uses a subset of the pandas offset aliases.
+
+        - 'Y', 'A' for yearly (long term average of 365.2425 days)
```
Comment (jreback) [inline pandas/core/tools/datetimes.py]: remove Y, A, M here I think we deprecated these anyhow

**Review by jreback — CHANGES_REQUESTED** [review]
we have a bunch of tests for this already in pandas/tests/indexes/timedeltas/test_tools.py

so I suppose some of these were not caught if you actually need to do this? 

also would need to do an asv run on appropriate benchmarks for this change

File: pandas/_libs/tslibs/timedeltas.pyx (hunk 1 above)
Comment (timcera) [reply pandas/_libs/tslibs/timedeltas.pyx]: I don't understand.  I actually want "unit" to be unchanged if not one of the tested values so what would an "else" do?  Wouldn't it force all "unit" values to be "U"?  I have in the past used a dictionary instead to do this same thing if that is a better and more readable approach.

**Comment by timcera** [comment]
> we have a bunch of tests for this already in pandas/tests/indexes/timedeltas/test_tools.py

No test for the added stride, also almost all of the tests are only for unit='D' and unit='s'.

> so I suppose some of these were not caught if you actually need to do this?

I think reasonable to add tests for stride and for other accepted units.  Working on it.

> also would need to do an asv run on appropriate benchmarks for this change

I have not done this before.  Will look around on how to make that happen.

File: pandas/core/tools/datetimes.py (hunk 2 above)
Comment (timcera) [reply pandas/core/tools/datetimes.py]: Already my working version is a deprecation for "Y", "A", "M" and "W".  The "W" deprecation is discussed earlier in this conversation.

File: pandas/_libs/tslibs/timedeltas.pyx (hunk 1 above)
Comment (jreback) [reply pandas/_libs/tslibs/timedeltas.pyx]:
```
else:
    unit = unit.upper()
```

File: pandas/core/tools/datetimes.py (hunk 2 above)
Comment (jreback) [reply pandas/core/tools/datetimes.py]: so don't mention the deprecated ones

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 3]
```diff
@@ -252,35 +254,54 @@ cpdef inline object precision_from_unit(object unit):
...
-    if unit == 'Y':
-        m = 1000000000L * 31556952
+    if unit is None:
+        m = 1L
+        p = 0
+        return m, p
+
+    unit, stride = _base_and_stride(unit)
+
+    # Normalize old or lowercase codes to standard offset aliases.
```
Comment (jreback) [inline pandas/_libs/tslibs/timedeltas.pyx]: I am not 100% sure we are actually testing all of these, can you see where we are and validate

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you merge master and respond to comments

**Comment by timcera** [comment]
Almost done.

Ran into this https://github.com/numpy/numpy/issues/8325, which affects several of the tests and for some time thought it was my code failing instead.  Tried to resolve/fix but just spun wheels instead.

Ran the benchmarks multiple times and got different answers each time.  For example, see the attached files of the results summaries from two benchmark runs.  Both done overnight while my Chromebook was not used for anything else.
[summary_results_2.txt](https://github.com/pandas-dev/pandas/files/3374138/summary_results_2.txt)
[summary_results_1.txt](https://github.com/pandas-dev/pandas/files/3374139/summary_results_1.txt)
Don't know what to do about that.  Copying pandas working directory to a beefier machine and see what happens.

Should I squash all commits into one?

**Comment by jreback** [comment]
> Almost done.
> 
> Ran into this [numpy/numpy#8325](https://github.com/numpy/numpy/issues/8325), which affects several of the tests and for some time thought it was my code failing instead. Tried to resolve/fix but just spun wheels instead.
> 
> Ran the benchmarks multiple times and got different answers each time. For example, see the attached files of the results summaries from two benchmark runs. Both done overnight while my Chromebook was not used for anything else.
> [summary_results_2.txt](https://github.com/pandas-dev/pandas/files/3374138/summary_results_2.txt)
> [summary_results_1.txt](https://github.com/pandas-dev/pandas/files/3374139/summary_results_1.txt)
> Don't know what to do about that. Copying pandas working directory to a beefier machine and see what happens.
> 
> Should I squash all commits into one?

we don't use numpy directly for (very little) nan operations, and not for timedeltas so not sure where you are hitting this.

**Comment by timcera** [comment]
> we don't use numpy directly for (very little) nan operations, and not for timedeltas so not sure where you are hitting this.

Initial bug report is https://github.com/pandas-dev/pandas/issues/26964 that shows the problem in pandas.  Closed it when I found that numpy had the same bug and thought, maybe incorrectly, that pandas was using numpy and therefore getting the same behavior.

**Comment by TomAugspurger** [comment]
Looks like a git merge issue. Let us know if you need help fixing.

**Comment by timcera** [comment]
Yep.  Need help with git.  I though everything looked good here, but obviously it is a mess.

I could send the log of what I had done if that is helpful.

**Comment by TomAugspurger** [comment]
In theory, a

```
git fetch upstream
git merge upstream/master
```

should get you straightened out.

**Comment by TomAugspurger** [comment]
numpydev is failing. Does this PR introduce new warnings from Numpy? Running with `pytest -wError` may help.

**Comment by jreback** [comment]
closing as stale; if you want to continue working, pls merge ``upstream/master`` and ping to re-open.

**Pull request closed without merging**